src/EventListener/CheckoutCartAmazonPayExpressEventListener.php

```diff
@@ -0,0 +1,163 @@
+use Shopware\Core\System\SalesChannel\Aggregate\SalesChannelPaymentMethod\SalesChannelPaymentMethodDefinition;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Shopware\Storefront\Page\Checkout\Cart\CheckoutCartPageLoadedEvent;
+use Shopware\Storefront\Page\Checkout\Confirm\CheckoutConfirmPageLoadedEvent;
+use Shopware\Storefront\Page\Checkout\Offcanvas\OffcanvasCartPageLoadedEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class CheckoutCartAmazonPayExpressEventListener implements EventSubscriberInterface
+{
+    public function __construct(
+        private readonly RequestParameterFactory $requestParameterFactory,
+        private readonly PayoneClientInterface $payoneClient,
+        private readonly LoggerInterface $logger,
+        private readonly ConfigReaderInterface $configReader,
+        private readonly DefaultPaymentFilterService $paymentFilterService,
+        private readonly CartExtensionService $cartExtensionService,
+        private readonly CartService $cartService,
+        private readonly Connection $connection,
+        private readonly ButtonConfiguration $buttonConfiguration
+    ) {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CheckoutCartPageLoadedEvent::class => 'onCartLoaded',
+            OffcanvasCartPageLoadedEvent::class => 'onCartLoaded',
+        ];
+    }
+
+    public function onCartLoaded(CheckoutCartPageLoadedEvent|OffcanvasCartPageLoadedEvent $event): void
+    {
+        $salesChannels = $this->connection->fetchFirstColumn(
+            sprintf(
+                'SELECT LOWER(HEX(assoc.`sales_channel_id`)) FROM `%s` AS assoc
+                    LEFT JOIN `%s` AS pm ON pm.`id` = assoc.`payment_method_id`
+                    WHERE assoc.`payment_method_id` = ? AND pm.`active` = 1',
+                SalesChannelPaymentMethodDefinition::ENTITY_NAME,
+                PaymentMethodDefinition::ENTITY_NAME
+            ),
+            [Uuid::fromHexToBytes(PayoneAmazonPayExpress::UUID)]
+        );
+
+        if (!\in_array($event->getSalesChannelContext()->getSalesChannelId(), $salesChannels, true)) {
+            // payment method is not assigned to sales-channel of is disabled.
+            return;
+        }
+
+        $filteredPaymentMethods = $this->paymentFilterService->filterPaymentMethods(
+            new PaymentMethodCollection([(new PaymentMethodEntity())->assign([
+                'id' => PayoneAmazonPayExpress::UUID,
+                'handlerIdentifier' => PayoneAmazonPayExpressPaymentHandler::class,
+            ])]),
+            new PaymentFilterContext(
+                salesChannelContext: $event->getSalesChannelContext(),
+                currency: $event->getSalesChannelContext()->getCurrency(),
+                cart: $event->getPage()->getCart(),
+                flags: [
+                    PaymentFilterContext::FLAG_SKIP_EC_REQUIRED_DATA_VALIDATION => true,
```

Comment:
As mentioned in my other comment, I would just add the flags here as values, not as keys.

Reply:
done